widget_driver/lib/src/dependency_provider.dart

```diff
@@ -0,0 +1,143 @@
+import 'utils/runtime_environment_info.dart';
+
+/// Use this class to create/resolve dependencies in your widgets.
+/// This class adds a safe way to create/resolve dependencies during testing where
+/// parent widgets are not forced to provide a mocked instance of your dependency.
+///
+/// The `DependencyProvider` will create the dependency object
+/// according to the provided builder when you are not running tests.
+///
+/// During testing the `DependencyProvider` will first try to create the dependency object using your builder,
+/// but if this fails and throws (because for example no parent widget registered the dependency),
+/// then the `DependencyProvider` will return the registered test default value.
+///
+/// This is how you use this in your widget where you provide dependencies:
+/// Step one: At the bottom of your widget class, create a subclass of `DependencyProvider` and implement the
+/// `registerTestDefaultFallbackValues` method. In this method you register the test default instance.
+/// This should be an empty implementation instance. Like a mock object.
+///
+/// ```dart
+/// class MyServiceProviderWidget extends StatelessWidget {
+///     ...
+/// }
+///
+/// class _DependencyProvider extends DependencyProvider {
+///     @override
+///     void registerTestDefaultFallbackValues() {
+///         registerTestDefaultBuilder<MyService>(() => TestDefaultMyService());
+///     }
+/// }
+///
+/// class TestDefaultMyService extends EmptyDefault implements TestDefaultMyService {}
+/// ```
+///
+/// And then you use this concrete version of `DependencyProvider` in your widget to create that instance.
+/// Like this:
+///
+/// ```dart
+/// class MyServiceProviderWidget extends StatelessWidget {
+///     ...
+///     @override
+///     Widget build(BuildContext context) {
+///         return Provider(
+///             create: _DependencyProvider().get(() => MyService(someStuffFromContext: context.read<SomeStuff>())),
+///             child: child,
+///         );
+///     }
+/// }
+/// ```
+///
+/// Now when you run this normally as a flutter app, then the real `MyService` will be created.
+/// And it will try to real `SomeStuff` out of the build context.
```

Comment:
aha.. yeah that might be what I meant :D 
But I think I just forgot to type some text.. I wanted to write:
`And it will try to get the real `SomeStuff` out of the build context.`

Reply:
Done